### ISSUE TITLE
OS X 10.9 (and earlier) support?

### ISSUE THREAD

**Issue by RJVB** [annulen/webkit#672]
Hi,

I understand this library is supposed to build to Qt versions as early as 5.6, but the (WebKit?) code currently imposes having at least v10.10 of the OS on Mac.

As far as I can tell at this point the few bits and pieces that are indeed dependent on the 10.10 or newer system SDKs are trivial to work around either by dropping an optional feature or by reintroducing some stripped code. Would you be interested in incorporating such changes, esp. if that would allow building the library on "endpoint" OS versions like 10.7? (I can no longer do test builds on pre-10.9 OS X set-ups myself but there should be a workaround for that too).

(10.7 is the latest version certain older but not yet obsolete Macs can run and for those it might be especially interesting to have access to an up-to-date browser framework)

**Comment by annulen** [comment]
Hi,

See https://github.com/annulen/webkit/issues/485#issuecomment-290676408 - it's said to work fine on 10.7 if built with 5.9 target

Note that if you are building full version with QML API enabled you'll probably need to switch IPC to Unix domain sockets, because Mac-specific code may break

if you want to restore something - well it depends on amount of changes, but it's probably fine if you can test this changes and they don't affect 10.10 builds

**Comment by RJVB** [comment]
> See https://github.com/annulen/webkit/issues/485#issuecomment-290676408 - it's said to work fine on 10.7 if built with 5.9 target

That must have been before the 10.10 requirement was introduced then.

I'll see about the IPC code. I'm currently rebuilding (yet again...) against the 10.9 SDK. That requires using the system malloc, in addition to deactivating the QoS tricks and telling CMake to accept pre-10.10 OS versions.

To 10.10SDK is a little annoying on 10.9 as it's not complete but still makes it look to code that it's being built on 10.10 . That's what makes the QoS deactivation necessary: the SDK has everything, but the runtime is lacking certain symbols.

Either way I'm setting up my adaptations properly ;)

**Comment by annulen** [comment]
>That must have been before the 10.10 requirement was introduced then

if you mean cmake condition, it was there (to make it clear that it's unsupported)

**Comment by annulen** [comment]
FWIW, in 4f5c3bc225bc884e7bc2f9a90a294c0d00b5df08 I've added a couple of options that are useful for you, you can build with -DUSE_UNIX_DOMAIN_SOCKETS=ON -DMACOS_USE_SYSTEM_ICU=OFF from now

**Comment by RJVB** [comment]
>FWIW, in 4f5c3bc225bc884e7bc2f9a90a294c0d00b5df08 I've added a couple of options that are useful for you, you can build with -DUSE_UNIX_DOMAIN_SOCKETS=ON -DMACOS_USE_SYSTEM_ICU=OFF from now

Thanks, how are these useful? What gets used instead of Unix domain sockets? And what if we want to use "our own" ICU libraries, like e.g. those in MacPorts? Will those still be picket up with MACOS_USE_SYSTEM_ICU=ON?

**Comment by annulen** [comment]
* By default mach ports are used on macOS, -DUSE_UNIX_DOMAIN_SOCKETS switches to unix sockets
* MACOS_USE_SYSTEM_ICU=ON means Apple ICU. Maybe name is not clear enough, but I think that system libraries on macOS are those shipped with system, not that user installed from 3rd party

**Comment by annulen** [comment]
So you need to disable it to use MacPorts ICU or whatever find_package finds

**Comment by RJVB** [comment]
>* By default mach ports are used on macOS, -DUSE_UNIX_DOMAIN_SOCKETS switches to unix sockets

No idea which is "better" but good to know the choice now exists.

>* MACOS_USE_SYSTEM_ICU=ON means Apple ICU. Maybe name is not clear enough, but I think that system libraries on macOS are those shipped with system, not that user installed from 3rd party

Actually, in Qt speak USE_SYSTEM_FOO usually means "don't use the copy we include but whatever you can find on the system given how the user set up the search paths". I just saw several examples of that in QtWebEngine.
If QtWebkit really interprets "MACOS_USE_SYSTEM" as "use the version installed in the system location" that should probably be documented clearly.

**Comment by annulen** [comment]
>No idea which is "better" but good to know the choice now exists.

This was discussed above:

>Note that if you are building full version with QML API enabled you'll probably need to switch IPC to Unix domain sockets, because Mac-specific code may break

Mach ports code paths requires 10.10 now and requirement may be raised after WebKit upgrade. Unix sockets may be slower (on macOS they are implemented on top of Mach ports internally) and lack QOS support, but API doesn't change

>Actually, in Qt speak USE_SYSTEM_FOO usually means "don't use the copy we include but whatever you can find on the system given how the user set up the search paths".

Ok. There is another option with similar naming: MACOS_FORCE_SYSTEM_XML_LIBRARIES. What names would you propose?

**Comment by RJVB** [comment]
>This was discussed above:

Sorry, that's the risk of replying via email without going to the ticket page and check if the answer was already there...

>Mach ports code paths requires 10.10 now

Sigh, there's clearly getting some more urgency to upgrading my OS. If only that were less disruptive (and likely to break my various workflows)...

>Ok. There is another option with similar naming: MACOS_FORCE_SYSTEM_XML_LIBRARIES. What names would you propose?

I'm drawing a blank. QWE now has configure commandline options `-qt-foo` to instruct it to use the internal copy; the same applies to e.g. libXCB in the QtBase configure. But that approach would probably oblige you to change your build system instead of just the option name. I think both naming approaches are justified:
- `-qt-foo` (_QT_FOO_LIBRARIES) means use the copy Qt includes
- `-system-foo` (_SYSTEM_FOO_LIBRARIES) means use the system copy, which in the strict sense of the word indeed means the copy in a system location.

The strict sense interpretation of SYSTEM would then make it impossible to build  against copies installed elsewhere (at least on Mac?).

In short, just document this Mac peculiarity in the option description and please make sure that those strict-on-Mac SYSTEM_FOO_LIBRARIES options all have a MACOS_ prefix. They can then simply be set to off in cross-platform packaging systems.
